File: Problems/Algorithms/1027. Longest Arithmetic Subsequence/LongestArithmeticSubsequence.java
class Solution {
    public int longestArithSeqLength(int[] nums) {
        int n = nums.length;
        HashMap<Integer, Integer>[] dp = new HashMap[nums.length];

        int ans = 0;
        for (int right = 0; right < nums.length; right++) {
            dp[right] = new HashMap<>();
            for (int left = 0; left < right; left++) {
                int difference = nums[right] - nums[left];
                dp[right].put(difference, dp[left].getOrDefault(difference, 1) + 1);
                ans = Math.max(ans, dp[right].get(difference));
            }
        }

        return ans;
    }
}

File: Problems/Algorithms/1218. Longest Arithmetic Subsequence of Given Difference/longest-arithmetic-subsequence.cpp
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    int longestSubsequence(vector<int>& arr, int difference) {
        unordered_map<int, int> dp;
        int ans = 1;

        for (int x : arr) {
            int last = dp.count(x - difference) ? dp[x - difference] : 0;
            dp[x] = last + 1;
            ans = max(ans, dp[x]);
        }

        return ans;
    }
};
